### PR TITLE
fix merging TargetClusters

### PR DESCRIPTION
**Pull request by yinsenyan**
same describe with #486 and fix it

### CODE REVIEW

File: pkg/scheduler/framework/interfaces/types.go [hunk 1]
```diff
@@ -137,6 +137,9 @@ func (t *TargetClusters) Merge(b *TargetClusters) {
 		}
 		for feed, bReplicas := range b.Replicas {
 			if len(bReplicas) != 0 {
+				if len(t.Replicas[feed]) == 0 {
```
Comment (Garrybest) [inline pkg/scheduler/framework/interfaces/types.go:140]: Which one is out of range, `t.Replicas[feed][ti]` or `bReplicas[bi]`?

Comment (Garrybest) [reply pkg/scheduler/framework/interfaces/types.go:140]: Could you please provide the panic log?

Comment (yinsenyan) [reply pkg/scheduler/framework/interfaces/types.go:140]: ``` t.Replicas[feed][ti] ``` panic with log : 
```
I0920 01:58:34.415392       1 dynamic.go:109] "Attempting to assign replicas to clusters" subscription="default/localstatus-divid-dynamic" clusters=[clusternet-svlrc/cls-xxxxxxx]
I0920 01:58:34.415452       1 trace.go:205] Trace[1284827553]: "Scheduling" namespace:default,name:localstatus-divid-dynamic (20-Sep-2022 01:58:34.219) (total time: 195ms):
Trace[1284827553]: ---"Computing predicates done" 0ms (01:58:34.219)
Trace[1284827553]: [195.635572ms] [195.635572ms] END
E0920 01:58:34.415537       1 runtime.go:78] Observed a panic: runtime.boundsError{x:0, y:0, signed:true, code:0x0} (runtime error: index out of range [0] with length 0)
goroutine 93 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x194f820, 0xc00072e720})
        /go/pkg/mod/github.com/clusternet/apimachinery@v0.23.0-alpha.0.0.20220224022903-dc3dec363e8c/pkg/util/runtime/runtime.go:74 +0x7d
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x11278a0})
        /go/pkg/mod/github.com/clusternet/apimachinery@v0.23.0-alpha.0.0.20220224022903-dc3dec363e8c/pkg/util/runtime/runtime.go:48 +0x75
panic({0x194f820, 0xc00072e720})
        /usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/clusternet/clusternet/pkg/scheduler/framework/interfaces.(*TargetClusters).Merge(0xc0004468c0, 0xc0006c5208)
        /go/src/github.com/clusternet/clusternet/pkg/scheduler/framework/interfaces/types.go:153 +0x8d2
github.com/clusternet/clusternet/pkg/scheduler/framework/plugins/defaultassigner.DynamicDivideReplicas(0xc0004c1800, {0xc000201200, 0x1, 0x13a5fd4}, {{0xc0005e8860, 0x0, 0x0}, 0xc00067a750})
        /go/src/github.com/clusternet/clusternet/pkg/scheduler/framework/plugins/defaultassigner/algorithm.go:52 +0x15b
github.com/clusternet/clusternet/pkg/scheduler/framework/plugins/defaultassigner.(*DynamicAssigner).Assign(0xe, {0xc, 0x3f08509d961747df}, 0xc0006c5498, 0xc0004c1800, 0xc000158000, {{0xc0005e8860, 0x1, 0x1}, 0xc00067a750})
        /go/src/github.com/clusternet/clusternet/pkg/scheduler/framework/plugins/defaultassigner/dynamic.go:119 +0x347
github.com/clusternet/clusternet/pkg/scheduler/framework/runtime.(*frameworkImpl).runAssignPlugin(0xc00066a960, {0x1cb0178, 0xc00044a940}, {0x7f9fbafe7ca8, 0xc0004cef10}, 0x1cb0178, 0xc00044a940, 0x7f9fbafe7c58, {{0xc0005e8860, 0x1, ...}, ...})
        /go/src/github.com/clusternet/clusternet/pkg/scheduler/framework/runtime/framework.go:611 +0xf2
github.com/clusternet/clusternet/pkg/scheduler/framework/runtime.(*frameworkImpl).RunAssignPlugins(0xc00066a960, {0x1cb0178, 0xc00044a940}, 0xb, 0x1658ca7, 0xa, {{0xc0005e8860, 0x1, 0x1}, 0xc00067a750})
        /go/src/github.com/clusternet/clusternet/pkg/scheduler/framework/runtime/framework.go:595 +0x26e
github.com/clusternet/clusternet/pkg/scheduler/algorithm.(*genericScheduler).selectClusters(0x1, {0x1cb0178, 0xc00044a940}, 0x1cfd630, {0xc000733d70, 0x1, 0xc0004c1800}, {0x1cfd630, 0xc00066a960}, 0xc0004c1800, ...)
        /go/src/github.com/clusternet/clusternet/pkg/scheduler/algorithm/generic.go:198 +0x39a
github.com/clusternet/clusternet/pkg/scheduler/algorithm.(*genericScheduler).Schedule(0xc000110c00, {0x1cb0178, 0xc00044a940}, {0x1cfd630, 0xc00066a960}, 0x0, 0xc0004c1800, 0xc0004a0e00)
        /go/src/github.com/clusternet/clusternet/pkg/scheduler/algorithm/generic.go:104 +0x58e
github.com/clusternet/clusternet/pkg/scheduler.(*Scheduler).scheduleOne(0xc000556000, {0x1cb0178, 0xc0001378c0})
        /go/src/github.com/clusternet/clusternet/pkg/scheduler/scheduler.go:281 +0x66c
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
        /go/pkg/mod/github.com/clusternet/apimachinery@v0.23.0-alpha.0.0.20220224022903-dc3dec363e8c/pkg/util/wait/wait.go:188 +0x25
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x7f9fbb1d91e0)
        /go/pkg/mod/github.com/clusternet/apimachinery@v0.23.0-alpha.0.0.20220224022903-dc3dec363e8c/pkg/util/wait/wait.go:155 +0x67
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x0, {0x1c7fa80, 0xc000732ea0}, 0x1, 0xc000758600)
        /go/pkg/mod/github.com/clusternet/apimachinery@v0.23.0-alpha.0.0.20220224022903-dc3dec363e8c/pkg/util/wait/wait.go:156 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x0, 0x0, 0x0, 0x0, 0x0)
        /go/pkg/mod/github.com/clusternet/apimachinery@v0.23.0-alpha.0.0.20220224022903-dc3dec363e8c/pkg/util/wait/wait.go:133 +0x89
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext({0x1cb0178, 0xc0001378c0}, 0xc000055fa0, 0x0, 0x0, 0x0)
        /go/pkg/mod/github.com/clusternet/apimachinery@v0.23.0-alpha.0.0.20220224022903-dc3dec363e8c/pkg/util/wait/wait.go:188 +0x99
k8s.io/apimachinery/pkg/util/wait.UntilWithContext(...)
        /go/pkg/mod/github.com/clusternet/apimachinery@v0.23.0-alpha.0.0.20220224022903-dc3dec363e8c/pkg/util/wait/wait.go:99
github.com/clusternet/clusternet/pkg/scheduler.(*Scheduler).Run.func1({0x1cb0178, 0xc0001378c0})
        /go/src/github.com/clusternet/clusternet/pkg/scheduler/scheduler.go:201 +0x4d
created by k8s.io/client-go/tools/leaderelection.(*LeaderElector).Run
        /go/pkg/mod/k8s.io/client-go@v0.23.1/tools/leaderelection/leaderelection.go:211 +0x154
panic: runtime error: index out of range [0] with length 0 [recovered]
        panic: runtime error: index out of range [0] with length 0

goroutine 93 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x11278a0})
        /go/pkg/mod/github.com/clusternet/apimachinery@v0.23.0-alpha.0.0.20220224022903-dc3dec363e8c/pkg/util/runtime/runtime.go:55 +0xd8
panic({0x194f820, 0xc00072e720})
        /usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/clusternet/clusternet/pkg/scheduler/framework/interfaces.(*TargetClusters).Merge(0xc0004468c0, 0xc0006c5208)
        /go/src/github.com/clusternet/clusternet/pkg/scheduler/framework/interfaces/types.go:153 +0x8d2
github.com/clusternet/clusternet/pkg/scheduler/framework/plugins/defaultassigner.DynamicDivideReplicas(0xc0004c1800, {0xc000201200, 0x1, 0x13a5fd4}, {{0xc0005e8860, 0x0, 0x0}, 0xc00067a750})
        /go/src/github.com/clusternet/clusternet/pkg/scheduler/framework/plugins/defaultassigner/algorithm.go:52 +0x15b
github.com/clusternet/clusternet/pkg/scheduler/framework/plugins/defaultassigner.(*DynamicAssigner).Assign(0xe, {0xc, 0x3f08509d961747df}, 0xc0006c5498, 0xc0004c1800, 0xc000158000, {{0xc0005e8860, 0x1, 0x1}, 0xc00067a750})
        /go/src/github.com/clusternet/clusternet/pkg/scheduler/framework/plugins/defaultassigner/dynamic.go:119 +0x347
github.com/clusternet/clusternet/pkg/scheduler/framework/runtime.(*frameworkImpl).runAssignPlugin(0xc00066a960, {0x1cb0178, 0xc00044a940}, {0x7f9fbafe7ca8, 0xc0004cef10}, 0x1cb0178, 0xc00044a940, 0x7f9fbafe7c58, {{0xc0005e8860, 0x1, ...}, ...})
        /go/src/github.com/clusternet/clusternet/pkg/scheduler/framework/runtime/framework.go:611 +0xf2
github.com/clusternet/clusternet/pkg/scheduler/framework/runtime.(*frameworkImpl).RunAssignPlugins(0xc00066a960, {0x1cb0178, 0xc00044a940}, 0xb, 0x1658ca7, 0xa, {{0xc0005e8860, 0x1, 0x1}, 0xc00067a750})
        /go/src/github.com/clusternet/clusternet/pkg/scheduler/framework/runtime/framework.go:595 +0x26e
github.com/clusternet/clusternet/pkg/scheduler/algorithm.(*genericScheduler).selectClusters(0x1, {0x1cb0178, 0xc00044a940}, 0x1cfd630, {0xc000733d70, 0x1, 0xc0004c1800}, {0x1cfd630, 0xc00066a960}, 0xc0004c1800, ...)
        /go/src/github.com/clusternet/clusternet/pkg/scheduler/algorithm/generic.go:198 +0x39a
github.com/clusternet/clusternet/pkg/scheduler/algorithm.(*genericScheduler).Schedule(0xc000110c00, {0x1cb0178, 0xc00044a940}, {0x1cfd630, 0xc00066a960}, 0x0, 0xc0004c1800, 0xc0004a0e00)
        /go/src/github.com/clusternet/clusternet/pkg/scheduler/algorithm/generic.go:104 +0x58e
github.com/clusternet/clusternet/pkg/scheduler.(*Scheduler).scheduleOne(0xc000556000, {0x1cb0178, 0xc0001378c0})
        /go/src/github.com/clusternet/clusternet/pkg/scheduler/scheduler.go:281 +0x66c
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
        /go/pkg/mod/github.com/clusternet/apimachinery@v0.23.0-alpha.0.0.20220224022903-dc3dec363e8c/pkg/util/wait/wait.go:188 +0x25
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x7f9fbb1d91e0)
        /go/pkg/mod/github.com/clusternet/apimachinery@v0.23.0-alpha.0.0.20220224022903-dc3dec363e8c/pkg/util/wait/wait.go:155 +0x67
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x0, {0x1c7fa80, 0xc000732ea0}, 0x1, 0xc000758600)
        /go/pkg/mod/github.com/clusternet/apimachinery@v0.23.0-alpha.0.0.20220224022903-dc3dec363e8c/pkg/util/wait/wait.go:156 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x0, 0x0, 0x0, 0x0, 0x0)
        /go/pkg/mod/github.com/clusternet/apimachinery@v0.23.0-alpha.0.0.20220224022903-dc3dec363e8c/pkg/util/wait/wait.go:133 +0x89
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext({0x1cb0178, 0xc0001378c0}, 0xc000055fa0, 0x0, 0x0, 0x0)
        /go/pkg/mod/github.com/clusternet/apimachinery@v0.23.0-alpha.0.0.20220224022903-dc3dec363e8c/pkg/util/wait/wait.go:188 +0x99
k8s.io/apimachinery/pkg/util/wait.UntilWithContext(...)
        /go/pkg/mod/github.com/clusternet/apimachinery@v0.23.0-alpha.0.0.20220224022903-dc3dec363e8c/pkg/util/wait/wait.go:99
github.com/clusternet/clusternet/pkg/scheduler.(*Scheduler).Run.func1({0x1cb0178, 0xc0001378c0})
        /go/src/github.com/clusternet/clusternet/pkg/scheduler/scheduler.go:201 +0x4d
created by k8s.io/client-go/tools/leaderelection.(*LeaderElector).Run
        /go/pkg/mod/k8s.io/client-go@v0.23.1/tools/leaderelection/leaderelection.go:211 +0x154
```

**Review by dixudx — APPROVED** [review]
I'm fine with this fix. Thanks for the contribution.

Leave to @Garrybest for another look.

File: pkg/scheduler/framework/interfaces/types.go (hunk 1 above)
Comment (Garrybest) [reply pkg/scheduler/framework/interfaces/types.go:140]: We should keep the slice for non-workload empty. So append 0 is not the best practice here. 

Here is my patch.

```go
func (t *TargetClusters) Merge(b *TargetClusters) {
	if b == nil || len(b.BindingClusters) == 0 {
		return
	}
	if t.Replicas == nil {
		t.Replicas = make(map[string][]int32)
	}
	// If the feed replicas does not exist in the former target clusters, we initialize them as an empty one.
	for feed, replicas := range b.Replicas {
		if _, exist := t.Replicas[feed]; exist {
			continue
		}
		if len(replicas) == 0 {
			t.Replicas[feed] = make([]int32, 0)
		} else if len(t.BindingClusters) != 0 {
			t.Replicas[feed] = make([]int32, len(t.BindingClusters))
		} else {
			t.Replicas[feed] = make([]int32, 0, len(replicas))
		}
	}

	// transfer from cluster to index
	m := make(map[string]int)
	for i, cluster := range t.BindingClusters {
		m[cluster] = i
	}

	for bi, bCluster := range b.BindingClusters {
		ti, exist := m[bCluster]
		if !exist {
			// this cluster is a new one, we should append
			t.BindingClusters = append(t.BindingClusters, bCluster)
			for feed, tReplicas := range t.Replicas {
				if cap(tReplicas) != 0 {
					t.Replicas[feed] = append(t.Replicas[feed], 0)
				}
			}
			ti = len(t.BindingClusters) - 1
			m[bCluster] = ti
		}
		for feed, bReplicas := range b.Replicas {
			if len(bReplicas) != 0 {
				t.Replicas[feed][ti] += bReplicas[bi]
			}
		}
	}
}
```

**Comment by Garrybest** [comment]
LGTM

**Review by dixudx — APPROVED** [review]

File: pkg/scheduler/framework/interfaces/types.go (hunk 1 above)
Comment (yinsenyan) [reply pkg/scheduler/framework/interfaces/types.go:144]: If we need append a bReplicas[bi] into t Replicas[feed],  we should not keep the slice for non-workload empty .
And use len(bReplicas)  when t bindingcluster is nil,  scheduler will panic with length of bindingcluster not equal length of some feed replicas.

**Comment by dixudx** [comment]
@yinsenyan https://github.com/gonum/gonum is a good library to handle matrix.

**Pull request merged**